### PR TITLE
Changes around navigator.webdriver

### LINKED ISSUE #922: Why Unforgeable
In chapter [4. Interface](https://w3c.github.io/webdriver/webdriver-spec.html#interface#h-interface):
> [Unforgeable]

I don't see how this makes sense, since navigator itself can be manipulated easily.

### LINKED ISSUE #923: Extend Navigator through a mixin
In chapter [4. Interface](https://w3c.github.io/webdriver/webdriver-spec.html#interface#h-interface):
> partial interface Navigator

The way you extend Navigator doesn't work for workers, which use WorkerNavigator. You need to define a [NoInterfaceObject] interface that is implemented by both Navigator and WorkerNavigator, similar to how the HTML Standard does it.

### LINKED ISSUE #947: Do not put navigator.webdriver on prototype chain
It was proposed in the [Intent To Ship](https://groups.google.com/a/chromium.org/d/msg/blink-dev/6GysDZCWwB8/U1RrBcsnBwAJ) of this feature in Chrome to expose `navigator.webdriver` only if UA is controlled by automation.

### PR DESCRIPTION
**Pull request by shekyan**
- expose `navigator.webdriver` if UA is controlled by WebDriver (fixes #947)
- drop [Unforgeable] (fixes #922)
- extend Navigator through mixin (fixes #923)

/cc @annevk, @foolip

---
This change is [Reviewable](https://reviewable.io/reviews/w3c/webdriver/948)

### CODE REVIEW

**Comment by shekyan** [comment]
I don't like how IDL is rendered, is there a way to avoid putting comments on new line?

**Comment by andreastt** [comment]
Paging @annevk for a sanity check on the WebIDL.

File: webdriver-spec.html [hunk 1]
```diff
@@ -699,13 +699,18 @@
...
 <h2>Interface</h2>
 
-<pre class=idl>partial interface Navigator {
-	[Unforgeable] readonly attribute boolean webdriver;
+<p>If the <dfn>webdriver-active flag</dfn> is true,
+ the user agent must support the following interface:
+
+<pre class=idl>
+	[NoInterfaceObject, Exposed=(Window,Worker)]
+	interface Navigator {
```
Comment (foolip) [inline webdriver-spec.html]: This should still be a partial interface.

File: webdriver-spec.html [hunk 2]
```diff
@@ -699,13 +699,18 @@
 <section>
 <h2>Interface</h2>
 
-<pre class=idl>partial interface Navigator {
-	[Unforgeable] readonly attribute boolean webdriver;
+<p>If the <dfn>webdriver-active flag</dfn> is true,
+ the user agent must support the following interface:
```
Comment (foolip) [inline webdriver-spec.html]: +partial

File: webdriver-spec.html [hunk 3]
```diff
@@ -699,13 +699,18 @@
 <section>
 <h2>Interface</h2>
 
-<pre class=idl>partial interface Navigator {
-	[Unforgeable] readonly attribute boolean webdriver;
+<p>If the <dfn>webdriver-active flag</dfn> is true,
```
Comment (foolip) [inline webdriver-spec.html]: Is it possible for this flag to change at runtime, i.e. is it possible to use WebDriver to connect to an already-running browser?

Comment (shekyan) [inline webdriver-spec.html]: As I understand from https://w3c.github.io/webdriver/webdriver-spec.html#new-session there is a hypothetical possibility to attach to a browser instance at runtime. 
> WebDriver implementations typically start a completely new browser instance, but there is no requirement in this specification (or for WebDriver only to be used to automate only web browsers). Implementations might choose to use an existing browser instance, eg. by selecting the window that currently has focus.

Implementations that I looked at (WebKit, Chrome) require new instance.

Comment (foolip) [inline webdriver-spec.html]: OK, well in that case I think this is fine, could tweak this in the future if needed.

**Comment by annevk** [comment]
This does not work. You need to define something similar to e.g. "NavigatorID" in HTML as you need to extend both `Navigator` and `WorkerNavigator`.

**Comment by shekyan** [comment]
@annevk , is this closer to the right way of doing it?

File: webdriver-spec.html [hunk 4]
```diff
@@ -699,13 +700,37 @@
...
-};</pre>
+<p>If the <dfn>webdriver-active flag</dfn> is true,
+ the user agent must support the following partial interfaces:
 
-<p data-dfn-for="Navigator">The <dfn><code>webdriver</code></dfn> IDL attribute
- of the <code><a>Navigator</a></code> interface must return
- the value of the <dfn>webdriver-active flag</dfn>, which is initially false.
+ <pre class=idl>
+ partial interface Navigator {
+ };
```
Comment (annevk) [inline webdriver-spec.html]: No need for this partial declaration.

File: webdriver-spec.html [hunk 5]
```diff
@@ -699,13 +700,37 @@
...
+ Navigator implements AutomationInformation;
+</pre>
+
+<pre class=idl>
+ [Exposed=Worker]
+ interface WorkerNavigator {};
```
Comment (annevk) [inline webdriver-spec.html]: No need for this interface declaration either.

File: webdriver-spec.html [hunk 6]
```diff
@@ -699,13 +700,37 @@
...
+ WorkerNavigator implements AutomationInformation;
+</pre>
+
+<section data-dfn-for="AutomationInformation" data-dfn-for="AutomationInformation">
+
+<h2><dfn>AutomationInformation</dfn> interface</h2>
```
Comment (annevk) [inline webdriver-spec.html]: It should have Navigator as a prefix. That's the pattern.

File: webdriver-spec.html [hunk 7]
```diff
@@ -699,13 +700,37 @@
...
+
+<pre class=idl>
+ [NoInterfaceObject, Exposed=(Window,Worker)]
+ interface AutomationInformation {
+ [Exposed=Window] readonly attribute boolean webdriver; // constant true
```
Comment (annevk) [inline webdriver-spec.html]: If you only expose it on Window, why extend both WorkerNavigator and Navigator?

Comment (shekyan) [inline webdriver-spec.html]: Uh, that's how I interpreted your comments in https://github.com/w3c/webdriver/issues/923 .

Comment (shekyan) [inline webdriver-spec.html]: @annevk, addressed comments.

File: webdriver-spec.html [hunk 8]
```diff
@@ -697,15 +698,26 @@
...
+ the user agent must support <dfn>NavigatorAutomationInformation</dfn> interface:
+
+ <pre class=idl>
+ Navigator implements NavigatorAutomationInformation;
+</pre>
+
+<pre class=idl>
+ [NoInterfaceObject, Exposed=(Window)]
+ interface NavigatorAutomationInformation {
+ [Exposed=Window] readonly attribute boolean webdriver; // constant true
```
Comment (annevk) [inline webdriver-spec.html]: You don't need the additional Exposed=Window line here. I'd make the comment "always returns true" as per precedent in DOM.

File: webdriver-spec.html [hunk 9]
```diff
@@ -697,15 +698,26 @@
...
+ };
+</pre>
+
+<p data-dfn-for="NavigatorAutomationInformation">Note that the <dfn>webdriver</dfn>
+ property should not be exposed on <code><a>WorkerNavigator</a></code>, and
+ its value must be true.
```
Comment (annevk) [inline webdriver-spec.html]: It's a getter that returns true. It doesn't have a value like that.

**Comment by annevk** [comment]
This looks good, though why does it always return true again? Wouldn't it normally return false?

**Comment by shekyan** [comment]
Oh, because we don't want to expose this property unless UA is controlled by webdriver. That's what Chrome folks suggested to do in my [Intent to ship](https://groups.google.com/a/chromium.org/d/msg/blink-dev/6GysDZCWwB8/Az0byBmIBwAJ)

**Comment by annevk** [comment]
I see, that kind of thing isn't really supported by IDL, but I guess if you address it in prose that should be fine. However, the IDL probably shouldn't suggest it always returns true then.

**Comment by shekyan** [comment]
I followed the pattern in https://html.spec.whatwg.org/multipage/system-state.html#client-identification
`If the navigator compatibility mode is Gecko, then the user agent must also support the following partial interface:`

```
partial interface NavigatorID {
  [Exposed=Window] boolean taintEnabled(); // constant false
  [Exposed=Window] readonly attribute DOMString oscpu;
};
```

**Comment by annevk** [comment]
Fair enough, all good then.

**Comment by shs96c** [comment]
This looks fine to me. Is this ready to be merged in?

**Comment by shekyan** [comment]
I don't have nothing more to add.

**Pull request merged**